syntax = "proto3";

package flwr.proto;

import "flwr/proto/fab.proto";
import "flwr/proto/run.proto";
import "flwr/proto/message.proto";

service ClientAppIo {
  // Get Message, Context, and Run
  rpc PullClientAppInputs(PullClientAppInputsRequest)
      returns (PullClientAppInputsResponse) {}

  // Send updated Message and Context
  rpc PushClientAppOutputs(PushClientAppOutputsRequest)
      returns (PushClientAppOutputsResponse) {}
}

enum ClientAppOutputCode {
  SUCCESS = 0;
  DEADLINE_EXCEEDED = 1;
  UNKNOWN_ERROR = 2;
}
message ClientAppOutputStatus {
  ClientAppOutputCode code = 1;
  string message = 2;
}

message PullClientAppInputsRequest { sint64 token = 1; }
message PullClientAppInputsResponse {
  Message message = 1;
  Context context = 2;
  Run run = 3;
}
message PushClientAppOutputsRequest {
  sint64 token = 1;
  Message message = 2;
  Context context = 3;
}
message PushClientAppOutputsResponse { ClientAppOutputStatus status = 1; }
